### PR TITLE
update doc to demonstrate usage of Persistence

### PR DESCRIPTION
**Pull request by bmahall**
This PR updates the persistence doc to address gaps in implementation of PersistenceLayer , particularly for Node-js desktop env.  An Electron sample containing full working code is being tracked as part of a separate work [item](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2138133/), however this doc includes the relevant code snippets with generic explanation wherever relevant.

### CODE REVIEW

File: extensions/docs/msal-node-extensions.md [hunk 1]
```diff
@@ -109,9 +166,61 @@ const publicClientConfig = {
...
+};
+
+```
+
 Note that MSAL will not read and write to persistence by default. You will have to call PublicClientApplication.tokenCache.readFromPersistence() and PublicClientApplication.tokenCache.writeToPersistence() anytime you trigger and MSAL Node operation that alters the token cache.
 
+
 ### Note for Electron Developers:
+Electron sample  link will be added here soon for clarity. 
+
```
Comment (bmahall) [inline extensions/docs/msal-node-extensions.md:255]: @derisen Thanks for your help in going over the documentation gaps and creating a separate work item to add the sample here ! Appreciate further feedback. 
CC: @salman90 @sameerag

**Review by peterzenz — APPROVED** [review]

**Comment by peterzenz** [comment]
@derisen could you review this one as well.  Thanks!

**Comment by bgavrilMS** [comment]
@Robbie-Microsoft - pls review as well.

File: extensions/docs/msal-node-extensions.md [hunk 2]
```diff
@@ -11,17 +11,49 @@ Supported platforms are Windows, Mac and Linux:
 - Linux - LibSecret is used for storing to "Secret Service" through npm keytar.
 
 ## Code
+
 ### Creating the persistence layer
-Creating the persistence will differ based on what platform you are targeting.
 
+API for creating the persistence layer will differ based on what platform you are targeting.
+
+Alternatively, you can use PersistenceCreator.createPersistence as it's a generic wrapper and selects the appropriate persistence based on the platform .
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
Alternatively, you can use the `createPersistence` API from [PersistenceCreator](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/msal-node-extensions/src/persistence/PersistenceCreator.ts) as it's a generic wrapper and selects the appropriate persistence method based on the platform/OS. 
```

File: extensions/docs/msal-node-extensions.md [hunk 3]
```diff
@@ -65,10 +105,17 @@ An unencrypted file persistence, which works across all platforms, is provided f
...
+const filePath = "path/to/cache/file.json";
+const filePersistence = FilePersistence.create(filePath, loggerOptions).then((persistence) => {
+            msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);
+
+            authProvider = new AuthProvider(msalConfig); //pass the msalConfig to AuthProvider constructor
+        });
+
 ```
 
+Note:- If file or directory has not been created, it FilePersistence.create() will create file and any directories in the path recursively.
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
> :information_source: If file or directory has not been created, `FilePersistence.create()` will create the file and any directories in the path recursively.
```

File: extensions/docs/msal-node-extensions.md (hunk 1 above)
Comment (derisen) [reply extensions/docs/msal-node-extensions.md:255]: Let's place [this link](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/electron-webpack) now, since we will eventually update it.

File: extensions/docs/msal-node-extensions.md [hunk 4]
```diff
@@ -51,7 +87,11 @@ import { LibSecretPersistence } from "@azure/msal-node-extensions";
 const cachePath = "path/to/cache/file.json";
 const serviceName = "";
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: Can we add example `serviceName `and `accountName `here, instead of empty strings?

File: extensions/docs/msal-node-extensions.md [hunk 5]
```diff
@@ -37,7 +69,11 @@ import { KeychainPersistence } from "@azure/msal-node-extensions";
...
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: Can we add example serviceName and accountName here, instead of empty strings?

File: extensions/docs/msal-node-extensions.md [hunk 6]
```diff
@@ -11,17 +11,49 @@ Supported platforms are Windows, Mac and Linux:
...
+Alternatively, you can use below platform-specific options:-
 #### Windows:
 ```js
 import { FilePersistenceWithDataProtection, DataProtectionScope } from "@azure/msal-node-extensions";
 
 const cachePath = "path/to/cache/file.json";
 const dataProtectionScope = DataProtectionScope.CurrentUser;
-const optinalEntropy = "";
-const windowsPersistence = FilePersistenceWithDataProtection.create(cachePath, dataProtectionScope, optionalEntropy);
+const optionalEntropy = "";
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: > const optionalEntropy = "";

Can we add a simple comment about what this does?

File: extensions/docs/msal-node-extensions.md [hunk 7]
```diff
@@ -87,11 +134,21 @@ const lockOptions = {
...
+
+            // Initialize the electron authenticator
+            authProvider = new AuthProvider(msalConfig);
+        });
+
 ```
 
 ### Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration
-Once you have a PersistenceCachePlugin, that can be set on the MSAL Node PublicClientApplication, by setting it as part of the configuration.
+Once you have a PersistenceCachePlugin, that can be set on the MSAL Node PublicClientApplication, by setting it as part of the configuration .
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
Once you have a `PersistenceCachePlugin`, that can be set on the MSAL Node `PublicClientApplication`, by setting it as part of the [configuration](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_node.html#configuration) object.
```

File: extensions/docs/msal-node-extensions.md [hunk 8]
```diff
@@ -109,9 +166,61 @@ const publicClientConfig = {
...
+module.exports = {
+  msalConfig: msalConfig,
+  protectedResources: protectedResources,
+  REDIRECT_URI: REDIRECT_URI,
+  persistenceConfiguration
+};
+
+```
+
 Note that MSAL will not read and write to persistence by default. You will have to call PublicClientApplication.tokenCache.readFromPersistence() and PublicClientApplication.tokenCache.writeToPersistence() anytime you trigger and MSAL Node operation that alters the token cache.
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: > Note that MSAL will not read and write to persistence by default. You will have to call PublicClientApplication.tokenCache.readFromPersistence() and PublicClientApplication.tokenCache.writeToPersistence() anytime you trigger and MSAL Node operation that alters the token cache.

Is this true? cc @sameerag 
I think cache read/write will be triggered whenever a public cache API is used e.g. `PublicClientApplication.tokenCache.getAllAccounts()` (not quite sure about acquireTokenSilent though)
Or am I missing the intent here?

File: extensions/docs/msal-node-extensions.md [hunk 9]
```diff
@@ -11,17 +11,49 @@ Supported platforms are Windows, Mac and Linux:
...
+authProvider = new AuthProvider(msalConfig);
+...
+
+PersistenceCreator
+        .createPersistence(persistenceConfiguration)
+        .then((persistence) => {
+            msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);
+
+            authProvider = new AuthProvider(msalConfig); //pass the msalConfig to AuthProvider constructor
+        });
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
const { PublicClientApplication } = require("@azure/msal-node");

const {
  DataProtectionScope,
  PersistenceCreator,
  PersistenceCachePlugin,
} = require("@azure/msal-node-extensions");

PersistenceCreator
        .createPersistence({
                cachePath: "path/to/cache/file.json",
                dataProtectionScope: DataProtectionScope.CurrentUser,
                serviceName: "test-msal-electron-service",
                accountName: "test-msal-electron-account",
                usePlaintextFileOnLinux: false,
          }).then((persistence) => {
          
            // use the persistence object to initalize an 
            // MSAL PublicClientApplication with cachePlugin
            const pca = new PublicClientApplication({
                    auth: {
                        clientId: "CLIENT_ID_HERE",
                    },
                    cache: {
                        cachePlugin: new PersistenceCachePlugin(persistence);
                    },
                });
        });
```

Comment (derisen) [reply extensions/docs/msal-node-extensions.md]: wanted to make this snippet more self contained (and generic), but feel free to edit! (I think we won't need the last snippet down below (authConfig.js) after this (actually the entire last section can go away)

File: extensions/docs/msal-node-extensions.md [hunk 10]
```diff
@@ -11,17 +11,49 @@ Supported platforms are Windows, Mac and Linux:
...
+  PersistenceCachePlugin,
+} = require("@azure/msal-node-extensions");
+
+let authProvider;
+authProvider = new AuthProvider(msalConfig);
+...
+
+PersistenceCreator
+        .createPersistence(persistenceConfiguration)
+        .then((persistence) => {
```
Comment (Robbie-Microsoft) [inline extensions/docs/msal-node-extensions.md]: Should we change this to: `const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration)`?

Should we be using await instead of .then here and everywhere else in this file?

File: extensions/docs/msal-node-extensions.md [hunk 11]
```diff
@@ -11,17 +11,49 @@ Supported platforms are Windows, Mac and Linux:
...
+            msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);
+
+            authProvider = new AuthProvider(msalConfig); //pass the msalConfig to AuthProvider constructor
+        });
+```
+
+Alternatively, you can use below platform-specific options:-
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
Alternatively, you can use below platform-specific options:
```

File: extensions/docs/msal-node-extensions.md (hunk 10 above)
Comment (derisen) [reply extensions/docs/msal-node-extensions.md]: +1

File: extensions/docs/msal-node-extensions.md (hunk 1 above)
Comment (derisen) [reply extensions/docs/msal-node-extensions.md:255]: Oh there's [this sample](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/msal-node-extensions) too -let's link both of them in this doc

Comment (bmahall) [reply extensions/docs/msal-node-extensions.md:255]: This https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/extensions/samples/electron-webpack is fairly self-sufficient, the second link seems related to this one. Should I add both here? @derisen

File: extensions/docs/msal-node-extensions.md [hunk 12]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
 
 const persistenceCachePlugin = new PersistenceCachePlugin(windowsPersistence); // or any of the other ones.
 ```
 
 To support concurrent access my multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.
 
 ```js
-import { PersistenceCachePlugin } from "@azure/msal-node-extensions";
+const {
+  PersistenceCreator
```
Comment (Robbie-Microsoft) [inline extensions/docs/msal-node-extensions.md]: nit: missing comma at the end of this line

File: extensions/docs/msal-node-extensions.md [hunk 13]
```diff
@@ -11,17 +11,64 @@ Supported platforms are Windows, Mac and Linux:
...
+// Use the persistence object to initialize an MSAL PublicClientApplication with cachePlugin
+const pca = new PublicClientApplication({
+                auth: {
+                        clientId: "CLIENT_ID_HERE",
+                    },
+                cache: {
+                        cachePlugin: new PersistenceCachePlugin(persistence);
+                    },
+                });
+```
```
Comment (Robbie-Microsoft) [inline extensions/docs/msal-node-extensions.md:45]: Did you add this much indentation on purpose? I'm wondering why you indented everything after the first line so much.

File: extensions/docs/msal-node-extensions.md [hunk 14]
```diff
@@ -11,17 +11,64 @@ Supported platforms are Windows, Mac and Linux:
...
+
+Alternatively, you can use below platform-specific options:
 #### Windows:
 ```js
-import { FilePersistenceWithDataProtection, DataProtectionScope } from "@azure/msal-node-extensions";
+
+const { FilePersistenceWithDataProtection, DataProtectionScope } = require("@azure/msal-node-extensions");
```
Comment (Robbie-Microsoft) [inline extensions/docs/msal-node-extensions.md:51]: Just curios: why did you change from import to require?

File: extensions/docs/msal-node-extensions.md (hunk 8 above)
Comment (sameerag) [reply extensions/docs/msal-node-extensions.md]: Not true. I think it should get triggered when we set `allowPeristence: true` in config and enable the wrapper on initial pca construction. Example [here](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/extensions/samples/msal-node-extensions/index.js).

**Review by sameerag — APPROVED** [review]
Address the usage issue highlighted by @derisen and I linked the sample. Please work with him for final approval. I am approving to unblock you. cc @derisen

File: extensions/docs/msal-node-extensions.md (hunk 1 above)
Comment (derisen) [reply extensions/docs/msal-node-extensions.md:255]: Wouldn't hurt I think (usually I like to end docs with a #samples section listing all the related samples)

File: extensions/docs/msal-node-extensions.md [hunk 15]
```diff
@@ -11,17 +11,64 @@ Supported platforms are Windows, Mac and Linux:
...
+
+const persistence = await PersistenceCreator.createPersistence({
+                cachePath: "path/to/cache/file.json",
+                dataProtectionScope: DataProtectionScope.CurrentUser,
+                serviceName: "test-msal-electron-service",
+                accountName: "test-msal-electron-account",
+                usePlaintextFileOnLinux: false,
+          });
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = new PersistenceCachePlugin(persistence);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
```

File: extensions/docs/msal-node-extensions.md [hunk 16]
```diff
@@ -11,17 +11,64 @@ Supported platforms are Windows, Mac and Linux:
...
+const { PublicClientApplication } = require("@azure/msal-node");
 
 const cachePath = "path/to/cache/file.json";
 const dataProtectionScope = DataProtectionScope.CurrentUser;
-const optinalEntropy = "";
-const windowsPersistence = FilePersistenceWithDataProtection.create(cachePath, dataProtectionScope, optionalEntropy);
+const optionalEntropy = ""; //specifies password or other additional entropy used to encrypt the data.
+const windowsPersistence = await FilePersistenceWithDataProtection.create(cachePath, dataProtectionScope, optionalEntropy);
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = new PersistenceCachePlugin(windowsPersistence);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
```

File: extensions/docs/msal-node-extensions.md [hunk 17]
```diff
@@ -32,12 +79,24 @@ The FilePersistenceWithDataProtection uses the Win32 CryptProtectData and CryptU
...
 
 const cachePath = "path/to/cache/file.json";
-const serviceName = "";
-const accountName = "";
-const macPersistence = KeychainPersistence.create(cachePath, serviceName, accountName);
+const serviceName = "test-msal-electron-service";
+const accountName = "test-msal-electron-account";
+const macPersistence = await KeychainPersistence.create(cachePath, serviceName, accountName);
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = new PersistenceCachePlugin(macPersistence);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
```

File: extensions/docs/msal-node-extensions.md [hunk 18]
```diff
@@ -46,12 +105,23 @@ const macPersistence = KeychainPersistence.create(cachePath, serviceName, accoun
...
 
 const cachePath = "path/to/cache/file.json";
-const serviceName = "";
-const accountName = "";
-const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, accountName);
+const serviceName = "test-msal-electron-service";
+const accountName = "test-msal-electron-account";
+const linuxPersistence = await LibSecretPersistence.create(cachePath, serviceName, accountName);
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = new PersistenceCachePlugin(linuxPersistence);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
```

File: extensions/docs/msal-node-extensions.md (hunk 15 above)
Comment (derisen) [reply extensions/docs/msal-node-extensions.md]: these lines are superfluous

File: extensions/docs/msal-node-extensions.md [hunk 19]
```diff
@@ -109,9 +196,62 @@ const publicClientConfig = {
...
+...
+
+module.exports = {
+  msalConfig: msalConfig,
+  protectedResources: protectedResources,
+  REDIRECT_URI: REDIRECT_URI,
+  persistenceConfiguration
+};
+
+```
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md:246]:
```suggestion
```

File: extensions/docs/msal-node-extensions.md [hunk 20]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
+const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration);
+const persistenceCachePlugin = new PersistenceCachePlugin(persistence, lockOptions); // or any of the other ones
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = persistenceCachePlugin;
+// Initialize the electron authenticator
+authProvider = new AuthProvider(msalConfig);
+
 ```
 
 ### Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: @bmahall let's remove this entire section (i.e. **###Setting the PersistenceCachePlugin on the MSAL Node PublicClientApplication configuration**) and everything under it, since in the snippets above we already show how to create a PublicClientApplication instance with the persistenceCachePlugin. For more details, we can just direct people to check the samples

File: extensions/docs/msal-node-extensions.md [hunk 21]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
 Create the PersistenceCachePlugin, by passing in the persistence object that was created in the previous step.
 
 ```js
-import { PersistenceCachePlugin } from "@azure/msal-node-extensions";
+const { PersistenceCachePlugin } = require("@azure/msal-node-extensions");
 
 const persistenceCachePlugin = new PersistenceCachePlugin(windowsPersistence); // or any of the other ones.
 ```
 
 To support concurrent access my multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
To support concurrent access by multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.
```

File: extensions/docs/msal-node-extensions.md [hunk 22]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
+ // Initialize the electron authenticator
+authProvider = new AuthProvider(msalConfig); 
 
-const cachePath = "path/to/cache/file.json";
-const filePersistence = FilePersistence.create(cachePath, serviceName, accountName);
 ```
 
+> :https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/e7ea9fd970c035aaee9f8d3c3d3196334fd1c0de/extensions/msal-node-extensions/src/persistence/FilePersistence.ts#L18: If file or directory has not been created, `FilePersistence.create()` will create the file and any directories in the path recursively.
+
 ### Creating the cache plugin
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]: > ### Creating the cache plugin

Can we retitle this section? We already covered how to create the cachePLuing in the snippets above, the crucial piece of information in this section is the part about this:

> To support concurrent access by multiple processess, the extensions use a file based lock. You can configure the retry number and retry delay for lock acquisition through CrossPlatformLockOptions.

File: extensions/docs/msal-node-extensions.md [hunk 23]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
     retryDelay: 50
 }
-const persistenceCachePlugin = new PersistenceCachePlugin(windowsPersistence, lockOptions); // or any of the other ones.
+
+const persistence = await PersistenceCreator.createPersistence(persistenceConfiguration);
+const persistenceCachePlugin = new PersistenceCachePlugin(persistence, lockOptions); // or any of the other ones
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = persistenceCachePlugin;
+// Initialize the electron authenticator
+authProvider = new AuthProvider(msalConfig);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
const pca = new PublicClientApplication({
    auth: {
            clientId: "CLIENT_ID_HERE",
        },
    cache: {
            cachePlugin: persistenceCachePlugin
        },
    });
```

File: extensions/docs/msal-node-extensions.md [hunk 24]
```diff
@@ -63,35 +133,52 @@ const linuxPersistence = LibSecretPersistence.create(cachePath, serviceName, acc
...
 ```js
-import { FilePersistence } from "@azure/msal-node-extensions";
+const { FilePersistence } = require("@azure/msal-node-extensions");
+
+const filePath = "path/to/cache/file.json";
+const filePersistence = await FilePersistence.create(filePath, loggerOptions);
+// Pass the persistence to msal config's cachePlugin
+msalConfig.cache.cachePlugin = new PersistenceCachePlugin(filePersistence);
+ // Initialize the electron authenticator
+authProvider = new AuthProvider(msalConfig);
```
Comment (derisen) [inline extensions/docs/msal-node-extensions.md]:
```suggestion
const pca = new PublicClientApplication({
    auth: {
            clientId: "CLIENT_ID_HERE",
        },
    cache: {
            cachePlugin: new PersistenceCachePlugin(filePersistence);
        },
  });
```

Comment (derisen) [reply extensions/docs/msal-node-extensions.md]: The reader doesn't know about our custom class `AuthProvider`, we should just instantiate `PublicClientApplication`

File: extensions/docs/msal-node-extensions.md (hunk 13 above)
Comment (bmahall) [reply extensions/docs/msal-node-extensions.md:45]: this is similar to the way we are using the config in the sample and for readability.

File: extensions/docs/msal-node-extensions.md (hunk 14 above)
Comment (bmahall) [reply extensions/docs/msal-node-extensions.md:51]: to make it consistent with js electron sample

File: extensions/docs/msal-node-extensions.md (hunk 17 above)
Comment (bmahall) [reply extensions/docs/msal-node-extensions.md]: addressed

File: extensions/docs/msal-node-extensions.md (hunk 8 above)
Comment (bmahall) [reply extensions/docs/msal-node-extensions.md]: removed

File: extensions/docs/msal-node-extensions.md (hunk 20 above)
Comment (bmahall) [reply extensions/docs/msal-node-extensions.md]: I have updated the title , would like to keep this for the sake of completeness.  However, if absolutely redundant happy to have a separate PR to remove it. @derisen
